### DISCUSSION TITLE
Using a library to parse config files

### DISCUSSION THREAD

**Discussion by TOMATO-ONE** [neutrinolabs/xrdp#2399] [Ideas]
Hi,all

I am using xrdp as RDP Proxy.
I have added a lot of settings for RDP Proxy in xrdp.ini.
I am modifying xrdp.ini when new xrdp is released.

Is it possible to put some of the `xrdp.ini` settings under `/etc/xrdp/conf.d/` and include them, similar to apache and nginx?

If that is possible, I could put the configuration regarding RDP Proxy in conf.d.
As another application, when adding the xorgxrdp rpm/deb package, We could add the configuration for Xorg in conf.d, which would make it easier for the package manager.

**Comment by metalefty** [comment]
I'm in positive position with that if it is not hard to implement and will not make code very complex. It is a good chance to add such feature because we're switching to v1.0.

**Comment by matt335672** [comment]
Sounds like a good idea.

We need to define behaviour in the event of colliding sections. So for example, if  we used `include` as a keyword and `/etc/xrdp/xrdp.ini` looked like this:-

```ini
[section 1]
aaa=bbb
ccc=ddd

include /etc/conf.d/xrdp.conf
```

and `/etc/conf.d/xrdp.conf` looked like this:-

```ini
[section 1]
ccc=eee
```

I'd expect the result to be either-

```ini
[section 1]
aaa=bbb
ccc=eee
```

or:-

```ini
[section 1]
ccc=eee
```

I could make a case for either one. What do people think?

**Reply by metalefty** [reply]
I'm not sure whether such a well-known conf.d structure has a common name.

My idea about such fragmented config files 
1. Do not use `include` keyword at all
1. `/etc/xrdp/xrdp.ini` is the primary config file
1. `/etc/xrdp/xrdp.ini.d` is the directory to put fragments
1. Load all files in `xrdp.ini.d` in dictionary order
1. Parameters loaded earlier will be overwritten by loaded later

### /etc/xrdp/xrdp.ini
```ini
[section 1]
aaa=bbb
ccc=ddd
```

### /etc/xrdp/xrdp.ini.d/001-foobar.ini
```ini
[section 1]
aaa=eee
fff=ggg
```

### /etc/xrdp/xrdp.ini.d/002-buzqux.ini
```ini
[section 1]
aaa=hhh
```


### Final result
```ini
[section 1]
aaa=hhh
ccc=ddd
fff=ggg
```

**Reply by TOMATO-ONE** [reply]
Thank you for taking it up.

I would like to be able to choose the include method of not overwriting as well.

Now xrdp doesn't do this, but here's an example.

**/etc/xrdp/xrdp.ini**
```ini
[section1]
aaa=bbb
ccc=ddd
```

**/etc/xrdp/xrdp.ini.d/xrdp-administrator-force-settings.ini**
```ini
[section1]
aaa=eee
```

**/home/user01/.xrdp.ini.d/user01-add-settings.ini**
```ini
[section1]
aaa=hhh
fff=ggg
```
**Final result**
```ini
[section1]
aaa=eee
ccc=ddd
fff=ggg
```
Note:This is intended to allow xrdp to display a customized login screen for each user once login authentication has been performed.

**Reply by matt335672** [reply]
Hi @TOMATO-ONE 

Accessing user-specific files can't in general be done from the xrdp daemon, even if it's running as root - to cater for things like NFS and Kerberos (or even SELinux), user files can only be read by a process running in the context of that user. At that point we need the user to be authenticated and authorized, and so we're beyond being able to use a login screen. So at the moment I can't see how we can make user-specific configs work.

**Reply by TOMATO-ONE** [reply]
Thank you. @matt335672 

I see, it seems that the xrdp architecture makes it difficult to achieve that.

>I would like to be able to choose the include method of not overwriting as well.

I withdraw the above request.

Let's prioritise being able to put the split configuration files under `/etc/xrdp/xrdp.ini.d/`.

**Reply by metalefty** [reply]
I don't understand the use case of user-specific xrdp.ini.

**Reply by TOMATO-ONE** [reply]
I wanted to change the IP address of the VNC/RDP Proxy destination for each user.
There may be a way to do this other than having user-specific xrdp.ini .

**Reply by matt335672** [reply]
Thanks for the info - interesting use-case.

So at the moment, you end up with a login screen with one session per user, and the user has to select the right session for them. Is that right?

What would you like the front end interface to look like?

**Reply by metalefty** [reply]
It is an interesting use case for me, too.  You don't need to withdraw the request so early but we can have a further discussion to find out better approaches.  I think it can be achieved by a different approach. It's just an idea, I came up with hostname substitution. 

In my idea, I would like to introduce some keywords for substitution. For example, if xrdp.ini is configured like this, `%u` will be replaced with the username. 

```
[vnc-proxy]
name=vnc-proxy
lib=libvnc.so
ip=%u.vnc.example.com.
port=ask5900
username=ask
password=ask
```

Configure DNS like this to sort connections to the backend.

```
meta.vnc.example.com IN A 10.59.0.1
matt.vnc.example.com IN A 10.59.0.2
tomato-one.vnc.example.com IN A 10.59.0.2
```

This is how admins define different proxy destinations per user.

**Reply by matt335672** [reply]
I like that, and what's more it's easy to do.

We've got a function `g_format_info_string()` which allows for strings to be parsed for exactly this purpose.

What do you think @TOMATO-ONE?

**Reply by TOMATO-ONE** [reply]
Thank you. @metalefty 

Great idea!
I see that there is also a way to put it in the `/etc/hosts` file on the VNC/RDP-Proxy host without having to rely on DNS.

By the way.
This topic is a derivative one, so I think it's better to start a separate discussion.
What do you think, @matt335672 , @metalefty ?

**Reply by metalefty** [reply]
Agreed.

**Reply by matt335672** [reply]
Could you raise an issue on this @TOMATO-ONE? We can then mark it as a feature request. I think we're agreed it's a useful thing to have.

I think your original thread has been somewhat hijacked - apologies. I'll change the title of this one.

**Reply by TOMATO-ONE** [reply]
I have created a new discussion #2410.
I will continue the discussion here.

@matt335672 
I will create a new issue later.

**Comment by metalefty** [comment]
Also, it is a good chance to bump `ini_version` for v1.0. It indicates config files need migration.

Furthermore, I would personally like to switch configuration file format to [TOML](https://en.wikipedia.org/wiki/TOML). INI is a mature config file format but AFAIK there's no specification. In contrast, TOML has a formal specification. Another advantage of TOML is the existence of a parser library. We don't need to implement our own config parser. 

Let's research about TOML parser library for C language available on each OS on this thread.

**Comment by matt335672** [comment]
I've had a quick look at TOML.

Initial comments:-
- Case-sensitive keys will cause a compatibility problem. The existing parser uses case insensitivity.
- include files are not supported directly in v1.0.0. I was hoping for something like that to implement this feature!
- It will make it much easier to read xrdp configs from other languages and utilities which could be useful.
- (edit) Only '#' is supported for comments

As regarding platform support, I can't find anything for C in FreeBSD ports. Debian 10 has support for Go, Rust, Perl, Python and Ruby but not C. CentOS 7 has Go and Python, but no C. Am I missing something?

**Reply by metalefty** [reply]
As far as I looked, very few C libraries for TOML are available on most platforms.  It seems that TOML is popular in script language but not in C language so far :(. [toml11](https://github.com/ToruNiina/toml11) is available on most platform but it is for C++.

What about keeping using INI but with a library instead? [iniparser](https://github.com/ndevilla/iniparser) is available on most platforms. Arch, CentOS, Debian and FreeBSD.

**Reply by matt335672** [reply]
**11:12 Updated **
I've had a look at libiniparser, and also [inih](https://github.com/benhoyt/inih) which is available on the same platforms.

This is a static analysis only. I've not tried implementing anything.

## Supported syntax
Both support ';' and '#' for comments.
- iniparser supports mixed-case keys. However, the case of the original key is lost. This is incompatible with bits of `xrdp.ini` where we use the case of the key for (e.g.) "Password=".
- iniparser only supports one key per section. Multiple lines with the same key are silently ignored. This is incompatible with the `param=` lines in `sesman.ini`.
- iniparser supports Y/y/T/t/1 for boolean true and N/n/F/f/0 for boolean false. This is slightly different from our `g_text2bool()` call.
- iniparser supports hexadecimal integers with an `0x` prefix and octal integers with an `0` prefix.
- inih supports all existing features, as it's so simple!


## Error reporting
iniparser uses a callback, whereas inih essentially just stops parsing with the error line returned.

## Maintenance
- Both are under permissive licenses which would be useful if we needed to embed one within xrdp for maintenance reasons.
- Both projects show some activity in the last year. You wouldn't expect a lot of activity on a project like this!
- iniparser has a more complete API. This means less code on our part, but also means more effort if the API changes. It hasn't for at least 5 years.

## Popularity
Popularity; both are used by a few other packages only in Debian11:-

```
mjb@debian11:~$ sudo apt-cache rdepends libiniparser1
libiniparser1
Reverse Depends:
  libiniparser-dev
  mtd-utils
  renderd
  cava
mjb@debian11:~$ sudo apt-cache rdepends libinih1
libinih1
Reverse Depends:
  libinih-dev
  xfsprogs
  libretro-mgba
  libmgba
  libinireader0
  gamemode-daemon
  imv
```

## Other information
I see that iniparser also comes with a rudimentary hash table which we could maybe use elsewhere. 

## Conclusion
Both projects are viable, compatible with xrdp, and look like they could be used to replace our existing parser, but could introduce incompatibilities with existing files.

Personally I'd prefer `iniparser` I think because of the poor error reporting in `inih`. It also provides more built-in functionality. The main worry I'd have with it is the handling of `param=` lines in `sesman.ini`, although we also need a way to handle the case of prompts on the xrdp login screen.

Thoughts?

**Reply by metalefty** [reply]
Thanks for the detailed research. One more thought regarding TOML. I think we can embed third-party library tomlc99 in xrdp source as a submodule.  As far as I searched on GitHub, I found many projects depend on tomlc99 and they embed it in their own source.

https://github.com/search?q=tomlc99+submodule&type=Code

**Comment by matt335672** [comment]
Personally I'd be happy with that approach. The licence looks OK. There's a couple of possible wrinkles:-
- We're insisting on a compiler that can handle C99, and probably the [switches from the Makefile](/cktan/tomlc99/blob/894902820a3ea2f1ec470cd7fe338bde54045cf5/Makefile#L8-L20). I'm happy with that, but Jay may not be - see for example #1722.
- We'll need to disable code formatting checks for that directory. That should be easily doable.
- We'll probably need to wrap the directory to interface with the Makefile it uses

It seems to me we now have three possible approaches to changing the parser.
- Use libh and stick with the existing formats. Advantages; no user action required, and least change. Disadvantages; we still have hard-to-parse config files. Error reporting is minimal.
- Use iniparser and refuse to accept ini files which haven't been updated. Advantages; Uses a tried and tested library. Files are easier to parse. Disadvantages; Some user impact which needs to be managed.
- Use toml99. Advantages; Easiest to parse. Disadvantages; Need to integrate with xrdp build and CI systems. C99 is forced on us. User impact is greatest - we may need to provide a migration utility.

Another thing to bear in mind is the original discussion point, which is to read the config from multiple files, if I'm correctly understanding it. This is somewhat at odds with making the configs easier to parse.

Is that a reasonable summary?

@jsorg71 - what do you think about this discussion so far?

**Reply by jsorg71** [reply]
@matt335672 I been assuming you guys are already defaulting to C99.
I don't have a strong opinion here.  I did try toml99 and iniparser and both seem to compile on everything I can think to try.

**Comment by matt335672** [comment]
@jsorg71 - we're currently on C90 (with some exceptions for your SPARC machine). POSIX (IEEE Std 1003.1-2017) mentions C99, so it would seem to be a good place to aim for. Our current [cppcheck CI test](https://github.com/neutrinolabs/xrdp/blob/791db0998a7ff338b0890a13a3d3fab80169ba1b/scripts/run_cppcheck.sh#L44-L47) uses C11.

Are you still using the SPARC machine for big-endian testing? There seems to be very little big-endian hardware out there these days.

**Comment by jsorg71** [comment]
@matt335672 This is getting a bit off topic but what exceptions for SPARC are you talking about?  Are you sure it's not more about an older GCC? 
It is true that there seems to be less big endian systems these days.  PowerPC can run big or little I think.  RISC-V I think is little endian only.  There is MIPS as well.  I think we should support big endian as long as major distros support.
What is the problem with supporting big endian?  I think we should start a new discussion if this important and keep this about the config library.

**Comment by matt335672** [comment]
Platform discussion moved to #2441
